pkg/webhook/pod/pods.go

```diff
@@ -0,0 +1,149 @@
+	operatorCli versioned.Interface
+	// pvc control
+	pvcControl controller.PVCControlInterface
+	// pd Control
+	pdControl pdapi.PDControlInterface
+}
+
+func NewPodAdmissionControl(kubeCli kubernetes.Interface, operatorCli versioned.Interface, PdControl pdapi.PDControlInterface, kubeInformerFactory kubeinformers.SharedInformerFactory, recorder record.EventRecorder) *PodAdmissionControl {
+	pvcInformer := kubeInformerFactory.Core().V1().PersistentVolumeClaims()
+	PVCControl := controller.NewRealPVCControl(kubeCli, recorder, pvcInformer.Lister())
+
+	return &PodAdmissionControl{
+		kubeCli:     kubeCli,
+		operatorCli: operatorCli,
+		pvcControl:  PVCControl,
+		pdControl:   PdControl,
+	}
+}
+
+func (pc *PodAdmissionControl) AdmitPods(ar v1beta1.AdmissionReview) *v1beta1.AdmissionResponse {
+
+	name := ar.Request.Name
+	namespace := ar.Request.Namespace
+	operation := ar.Request.Operation
+	glog.Infof("receive admission to %s pod[%s/%s]", operation, namespace, name)
+
+	switch operation {
+	case v1beta1.Delete:
+		return pc.admitDeletePods(name, namespace)
+	default:
+		glog.Infof("Admit to %s pod[%s/%s]", operation, namespace, name)
+		return util.ARSuccess()
+	}
+}
+
+// Webhook server receive request to delete pod
+// if this pod wasn't member of tidbcluster, just let the request pass.
+// Otherwise, we check tidbcluster and statefulset which own this pod whether they were existed.
+// If either of them were deleted,we would let this request pass,
+//// otherwise we will check it decided by component type.
+func (pc *PodAdmissionControl) admitDeletePods(name, namespace string) *v1beta1.AdmissionResponse {
+
+	pod, err := pc.kubeCli.CoreV1().Pods(namespace).Get(name, metav1.GetOptions{})
+	if err != nil {
+		glog.Infof("failed to find pod[%s/%s] during delete it,admit to delete", namespace, name)
+		return util.ARSuccess()
+	}
+
+	l := label.Label(pod.Labels)
+	if !(l.IsPD() || l.IsTiKV() || l.IsTiDB()) {
+		glog.Infof("pod[%s/%s] is not TiDB component,admit to delete", namespace, name)
+		return util.ARSuccess()
+	}
+
+	tcName, exist := pod.Labels[label.InstanceLabelKey]
+	if !exist {
+		glog.Errorf("pod[%s/%s] has no label: %s", namespace, name, label.InstanceLabelKey)
+		return util.ARFail(fmt.Errorf("pod[%s/%s] has no label: %s", namespace, name, label.InstanceLabelKey))
+	}
+
+	tc, err := pc.operatorCli.PingcapV1alpha1().TidbClusters(namespace).Get(tcName, metav1.GetOptions{})
+
+	if err != nil {
+		if errors.IsNotFound(err) {
+			glog.Infof("tc[%s/%s] had been deleted,admit to delete pod[%s/%s]", namespace, tcName, namespace, name)
+			return util.ARSuccess()
+		}
+		glog.Errorf("failed get tc[%s/%s],refuse to delete pod[%s/%s]", namespace, tcName, namespace, name)
+		return util.ARFail(err)
+	}
+
+	if len(pod.OwnerReferences) == 0 {
+		return util.ARSuccess()
+	}
+
+	ownerStatefulSetName := pod.OwnerReferences[0].Name
+	ownerStatefulSet, err := pc.kubeCli.AppsV1().StatefulSets(namespace).Get(ownerStatefulSetName, metav1.GetOptions{})
```

Comment:
in theory, it is possible that a pod may be owned by more than one objects, I think we should find in the list of owner references and check if the pod is owned by a StatefulSet and then return the StatefulSet name.

Reply:
fix

Reply:
Use lister for sts?